### PR TITLE
Add keybind for key release events

### PR DESCRIPTION
**Pull request by t-wissmann**
This follows the suggestion of #1443 to implement release keybinds using
a fake modifier `Release`.

There are still some member variables that should be renamed. E.g. `grabbed`
should be `active`, but I didn't rename it because the diff is already quite
long.

Thanks to @madroach for suggesting things and for providing the first
implementation.

### CODE REVIEW

**Comment by t-wissmann** [comment]
The failing test case comes from mouse binds which also use the modifier list, but don't support bind on release. So a solution needs to be found. Either we let mouse and key binds have separate modifier lists, or we add a `--release` switch to `keybind`, or we have another level of modifiers, something like `Release:Mod4+s`.

**Comment by t-wissmann** [comment]
@madroach please see also the new test cases for this: https://github.com/herbstluftwm/herbstluftwm/pull/1446/commits/b6996635d3a31dedc17005ccca6343875fcbab16

**Comment by madroach** [comment]
#### Behaviour

Hi, I just tested this PR. It works as described, but still does not fulfill my "simple" wish:

>  The behaviour I want is this: The panel is shown only while the $Mod key is down.

> 1. what keys do you press/release

```
hc keybind Super_L         emit_hook unhidepanel
hc keybind Release+Super_L emit_hook hidepanel
```
Press and hold Ctrl
Press and hold Super_L
Release Super_L
Release Ctrl

> 2. what happens

Nothing

> 3. what you would expect hlwm to do instead

The panel is shown only while the Super_L key is pressed.

#### Implementation

To me it seems the implementation is needlessly complex. Tracking the modifier state of key presses and matching them to the release event is needless complexity just to avoid using `AnyModifier`.
It could be implemented without tracking the modifiers and without counting users of KeyGrabs, just by scanning the existing bindings.
If you are interested in a simple implementation I could provide one.

**Comment by t-wissmann** [comment]
> #### Behaviour
> Hi, I just tested this PR. It works as described, but still does not fulfill my "simple" wish:
> 
> > The behaviour I want is this: The panel is shown only while the $Mod key is down.

> > 1. what keys do you press/release
> 
> ```
> hc keybind Super_L         emit_hook unhidepanel
> hc keybind Release+Super_L emit_hook hidepanel
> ```
> 
> Press and hold Ctrl Press and hold Super_L Release Super_L Release Ctrl
> 
> > 2. what happens
> 
> Nothing

It is a basic principle of hlwm and computer keyboards in general, that if you bind
`Super_L`, then this key binding is *not* activated if you press
`Ctrl-Super_L`. And this observation is **entirely independent** of panels, or
release events. And it is **very good** that it is this way, because many
thing would break otherwise. Just think of a terminal in which you can press
`Ctrl-Shift-C` to copy text to the clipboard. With your logic, pressing
`Ctrl-Shift-C` should also activate the key binding `Ctrl-C` which terminates
the process running in the terminal.

So what you are actually asking for is two things: 1. key release events, 2. a
<kbd>Any</kbd> modifier. These are different. Already now, you can emulate the anymodifier via:
```
for mods in {Shift-,}{Ctrl-,}{Alt-,} ; do
    hc keybind ${mods}Super_L         emit_hook unhidepanel
    hc keybind ${mods}Release-Super_L emit_hook hidepanel
done
```
I don't object implementing an `Any-` modifier in hlwm, but I am strictly against breaking the `keybind` semantics.


> #### Implementation
> To me it seems the implementation is needlessly complex. Tracking the
> modifier state of key presses and matching them to the release event is
> needless complexity just to avoid using `AnyModifier`. It could be
> implemented without tracking the modifiers and without counting users of
> KeyGrabs, just by scanning the existing bindings. If you are interested in a
> simple implementation I could provide one.

I will not break the keybind implementation just to implement this one use-case
of hiding the panel while Super_L is pressed. You did not show me a single implementation of the "key release" feature that keeps the interface of the `keybind` and `keyunbind` commands sane.

**Comment by madroach** [comment]
Here's my implementation of `Release` and `Any` modifier: https://github.com/herbstluftwm/herbstluftwm/compare/master...madroach:bind_KeyRelease.

**Review by blarz — COMMENTED** [review]
Just two minor issues, otherwise looks good. Love the amount of tests added!

File: tests/conftest.py [hunk 1]
```diff
@@ -1051,6 +1051,7 @@ def keyboard():
     class KeyBoard:
         def press(self, key_spec):
             subprocess.check_call(['xdotool', 'key', key_spec])
+            print(f"sended key press {key_spec}", file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (t-wissmann) [reply tests/conftest.py]: I can't remember :-) But maybe it is helpful in the test suite logs. I've just moved the print before the subprocess call and replaced the anyway-wrong "sended" with "send".

File: src/xmainloop.cpp [hunk 2]
```diff
@@ -532,8 +533,8 @@ void XMainLoop::focusin(XFocusChangeEvent* event) {
     duringFocusIn_ = false;
 }
 
-void XMainLoop::keypress(XKeyEvent* event) {
-    //HSDebug("name is: KeyPress\n");
+void XMainLoop::keypressOrRelease(XKeyEvent* event) {
+    HSDebug("name is: KeyPress\n");
```
[the remark this answers is not in the document]
Comment (t-wissmann) [reply src/xmainloop.cpp]: Yes, I've updated the message and moved it back into the comment.

File: tests/conftest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (blarz) [reply tests/conftest.py]: Thanks!

File: src/xmainloop.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (blarz) [reply src/xmainloop.cpp]: Nice, thanks!

**Review by blarz — APPROVED** [review]

**Pull request merged**